File: 476.数字的补数.js
/*
 * @lc app=leetcode.cn id=476 lang=javascript
 *
 * [476] 数字的补数
 */

// @lc code=start
/**
 * @param {number} num
 * @return {number}
 */
 var findComplement = function(num) {
    return (1 << (32 - Math.clz32(num))) - 1 - num;
}
// var findComplement = function(num) {
//     let tmp = num;
//     let i = 0;
//     const arr = [];
//     do {
//         i = tmp & 1;
//         tmp = tmp >> 1;
//         arr.unshift(i);
//     } while (tmp !== 0);
//     let ret = 0;
//     for (let i = 0; i < arr.length; i++) {
//         ret = ret << 1;
//         ret += (1 - arr[i]);
//     }
//     return ret;
// };
// @lc code=end

// var findComplement = function(num) {
//   return parseInt(num.toString(2).split('').map(item => Number(!Number(item))).join(''), 2)
// };

var findComplement = function(num) {
    return 1 << (32 - Math.clz32(num)) - 1 - num;
}

File: 476.数字的补数.rs
/*
 * @lc app=leetcode.cn id=476 lang=rust
 *
 * [476] 数字的补数
 */

// @lc code=start
impl Solution {
    pub fn find_complement(num: i32) -> i32 {
        // 当 num = 2147483647，即 2 ^ 31 - 1，这是最大的32位带符号正整数。
        // 1 左移 31 位，应当为正数。如果用 i32 表示，则为负数，不满足题目要求。
        // (1 << (32 - num.leading_zeros())) as i32 - 1 - num
        // 全覆盖 1，减去 num
        ((1 << (32 - num.leading_zeros())) as u32 - 1 - (num as u32)) as i32

    }
}
// @lc code=end

